src/v/coproc/router.cc

```diff
@@ -320,49 +336,21 @@ ss::future<storage::log> router::get_log(const model::ntp& ntp) {
       storage::ntp_config(ntp, _api.local().log_mgr().config().base_dir));
 }
 
-errc router::add_source(
+ss::future<errc> router::add_source(
   const script_id id,
   const model::topic_namespace& tns,
   topic_ingestion_policy p) {
-    // For now only support the 'latest' policy
-    if (!is_valid_ingestion_policy(p)) {
-        return errc::invalid_ingestion_policy;
-    }
-    auto logs = _api.local().log_mgr().get(tns);
-    if (logs.empty()) {
-        return errc::topic_does_not_exist;
-    }
-
-    for (auto& [ntp, log] : logs) {
-        auto found = _sources.find(ntp);
-        if (found == _sources.end()) {
-            auto ts = ss::make_lw_shared<topic_state>(topic_state{
-              .log = log, .head = topic_offsets(), .scripts = {id}});
-            _sources.emplace(ntp, ts);
-        } else {
-            found->second->scripts.emplace(id);
-        }
-        vlog(coproclog.info, "Inserted ntp {} id {}", ntp, id);
-    }
-    return errc::success;
+    source<errc> s{.tn = tns, .tip = p};
+    ss::future<errc> f = s.promise.get_future();
+    _deferred_additions.emplace(id, std::move(s));
+    return f;
 }
 
-bool router::remove_source(const script_id sid) {
-    absl::flat_hash_set<model::ntp> deleted;
-    std::for_each(_sources.begin(), _sources.end(), [&deleted, sid](auto& p) {
-        auto& scripts = p.second->scripts;
-        scripts.erase(sid);
-        vlog(coproclog.info, "Deleted script id: {}", sid);
-        if (scripts.empty()) {
-            deleted.emplace(p.first);
-        }
-    });
-    // If no more scripts are tracking an ntp, remove the ntp
-    absl::erase_if(_sources, [&deleted](const auto& p) {
-        return deleted.contains(p.first);
-    });
-
-    return !deleted.empty();
+ss::future<bool> router::remove_source(const script_id sid) {
+    ss::promise<bool> p;
+    ss::future<bool> f = p.get_future();
+    _deferred_removals.emplace(sid, std::move(p));
```

Comment:
should we assert here that `emplace` succeeded?

Reply:
this is resolved but doesn't look like anything change or any follow conversation

Reply:
was already modified in a later commit

Reply:
i see it now.